File: BOJ/GiSeok/Python/BOJ_2206.py
# 백준_2206 - 벽 부수고 이동하기[그래프] (08/09)
import sys, copy
from collections import deque
input = sys.stdin.readline

N, M = map(int, input().split())
path = [list(map(int, input().rstrip())) for _ in range(N)]
visited =[[[0, 0] for i in range(M)] for _ in range(N)]
wasd = ((0, -1), (0, 1), (-1, 0), (1, 0))

def bfs(start_x, start_y):
    queue = deque()
    queue.append([start_x, start_y, 0])

    visited[start_y][start_x][0] = True
    dist = copy.deepcopy(path)
    dist[start_y][start_x] = 1

    while queue:
        x, y, wall_break = queue.popleft()
        
        if x == M-1 and y == N-1:
            return dist[y][x]

        for xx, yy in wasd:
            xx += x
            yy += y

            if (0 <= xx < M) and (0 <= yy < N):
                if path[yy][xx] == 0:
                    if not visited[yy][xx][wall_break]:
                        queue.append([xx, yy, wall_break])
                        visited[yy][xx][wall_break] = True
                        dist[yy][xx] = dist[y][x] + 1
                elif path[yy][xx] == 1 and wall_break == 0:
                    if not visited[yy][xx][0]:
                        queue.append([xx, yy, wall_break + 1])
                        visited[yy][xx][0] = True
                        dist[yy][xx] = dist[y][x] + 1
    
    return -1

print(bfs(0, 0))

File: BOJ/GiSeok/C++/BOJ_2206.cpp
#include <iostream>
#include <vector>
#include <queue>

struct point
{
    int x, y;
    int wallBreak;
};

int map[1001][1001];
int c_map[1001][1001];
bool visited[1001][1001][2];
std::vector<point> wall;
int N, M;

int bfs(point P)
{
    int wasd[4][2] = { {0, -1}, {0, 1}, {-1, 0}, {1, 0} };
    
    std::queue<point> q;
    q.push(P);
    visited[P.y][P.x][0] = true;
    
    for (int i = 1; i <= N; i++) {
        for (int j = 1; j <= M; j++) {
            c_map[i][j] = map[i][j];
        }
    }
    c_map[P.y][P.x] = 1;
    
    while (!q.empty())
    {
        point p = q.front();
        q.pop();
        
        if (p.x == M && p.y == N)
            return c_map[p.y][p.x];
 
        for (int i = 0; i < 4; i++) {
            if ((p.x + wasd[i][0] >= 1 && p.x + wasd[i][0] <= M) && (p.y + wasd[i][1] >= 1 && p.y + wasd[i][1] <= N)) {
                if (map[p.y + wasd[i][1]][p.x + wasd[i][0]] == 0) {
                    if (!visited[p.y + wasd[i][1]][p.x + wasd[i][0]][p.wallBreak]) {
                        q.push(point{p.x + wasd[i][0], p.y + wasd[i][1], p.wallBreak});
                        c_map[p.y + wasd[i][1]][p.x + wasd[i][0]] = c_map[p.y][p.x] + 1;
                        visited[p.y + wasd[i][1]][p.x + wasd[i][0]][p.wallBreak] = true;
                    }
                } else if (map[p.y + wasd[i][1]][p.x + wasd[i][0]] == 1 && p.wallBreak == 0) {
                    q.push(point{p.x + wasd[i][0], p.y + wasd[i][1], 1});
                    c_map[p.y + wasd[i][1]][p.x + wasd[i][0]] = c_map[p.y][p.x] + 1;
                }
            }
        }
    }
    
    return -1;
}

int main()
{
    std::cin >> N >> M;
    
    for (int i = 1; i <= N; i++) {
        for (int j = 1; j <= M; j++) {
            scanf("%1d", &map[i][j]);
        }
    }
    
    std::cout << bfs(point{1, 1, 0}) << std::endl;
}
